Sources/third_party/Experimental/Complex.swift

```diff
@@ -0,0 +1,363 @@
+// limitations under the License.
+//
+//  Note
+//  ====
+//
+//  For maximum consistency with corresponding functions in C/C++, checks for
+//  special values in `naturalExponential()`, `squareRoot()`, trigonometric
+//  functions, and hyperbolic functions are adapted from libc++.
+//
+//  Code in libc++ is dual-licensed under the MIT and UIUC/NCSA licenses.
+//  Copyright © 2009-2017 contributors to the LLVM/libc++ project.
+/// A type to represent a complex value in Cartesian form.
+///
+/// - Note: `Complex64` is a type alias for `Complex<Float>` and `Complex128` is
+///   a type alias for `Complex<Double>`.
+///
+/// Create new instances of `Complex<T>` using integer or floating-point
+/// literals and the imaginary unit `Complex<T>.i`. For example:
+///
+/// ```swift
+/// let x = 2 + 4 * .i // `x` is of type `Complex<Double>`
+/// let y = 3.5 + 7 * .i // `y` is of type `Complex<Double>`
+///
+/// let z: Complex64 = .e + .pi * .i // `z` is of type `Complex<Float>`
+/// ```
+///
+/// Additional Considerations
+/// -------------------------
+///
+/// Floating-point types have special values that represent infinity or NaN
+/// ("not a number"). Complex functions in different languages may return
+/// different results when working with special values.
+///
+/// Many complex functions have [branch cuts][dfn], which are curves in the
+/// complex plane across which a function is discontinuous. Different languages
+/// may adopt different branch cut structures for the same complex function.
+///
+/// Implementations in `Complex<T>` adhere to the [C standard][std] (Annex G) as
+/// closely as possible with respect to special values and branch cuts.
+///
+/// To users unfamiliar with complex functions, the principal value returned by
+/// some complex functions may be unexpected. For example,
+/// `Double.cbrt(-8) == -2`, which is the __real root__, while
+/// `Complex.cbrt(-8) == 2 * Complex.exp(.i * .pi / 3)`, which is the
+/// __principal root__.
+///
+/// [dfn]: http://mathworld.wolfram.com/BranchCut.html
+/// [std]: http://www.open-std.org/JTC1/SC22/WG14/www/standards.html#9899
+
+struct Complex<T: FloatingPoint> {
+    var real: T
+    var imaginary: T
+
+    @differentiable(vjp: _vjpInit where T: Differentiable, T.TangentVector == T)
+    init(real: T = 0, imaginary: T = 0) {
+        self.real = real
+        self.imaginary = imaginary
+    }
+}
+
+extension Complex: Differentiable where T: Differentiable {
+    typealias TangentVector = Complex
+    typealias AllDifferentiableVariables = Complex
+}
+
+extension Complex {
+    static var i: Complex {
+        return Complex(real: 0, imaginary: 1)
+    }
+
+    var isFinite: Bool {
+        return real.isFinite && imaginary.isFinite
+    }
+
+    var isInfinite: Bool {
+        return real.isInfinite || imaginary.isInfinite
+    }
+
+    var isNaN: Bool {
+        return (real.isNaN && !imaginary.isInfinite) ||
+        (imaginary.isNaN && !real.isInfinite)
```

Comment:
```suggestion
            (imaginary.isNaN && !real.isInfinite)
```

Reply:
I can actually even make this one line, so I'll do that